Implement sending via sidecar

bwoebi opened the pull request:
# What does this PR do?

* Endpoint is exposed via FFI
* Sidecar gains an API to handle agent sampling rates independently of the backend (i.e. working with sidecar and without sidecar). This is handled via mapped (shared) memory pages.
  * The goal is to have an _as up-to-date version as possible_, but be lock-free. This is achieved by just ignoring the shared data while writes to the shared memory are happening, falling back to a prior process-local copy.
* Sidecar config is now targets an Endpoint, which can be dynamically manipulated to fit the target url, with a possible API key. (telemetry mock file targets file:// Endpoint)
* Sidecar uses trace-* crates to send data to an agent or intake endpoint.
* Some common paths between sidecar and trace-mini-agent have been factored out into trace-utils. @DataDog/libdatadog-serverless, please review that part.
* File passing in ipc got a macro handling the boilerplate code for FileTransferHandles.
  * Primary motivation is that I've been utterly confused how this works (I was expecting some magic recognizing a trait on some args or such alongside serde) until after a while I discovered that explicit `impl TransferHandles for ExampleInterfaceRequest`, for both, which is just trivial boilerplate. The macro makes it explicit in the service definition what is passed.
* Global configuration is now directly passed via env upon sidecar creation. (guarding the self-telemetry for example)

The changes to the sidecar crate have been tested and validated in the dd-trace-php repository.

I'm aware the PR is quite big, but everything needed to come together to do proper end-to-end testing.

thedavl commented on trace-utils/src/trace_utils.rs:
```diff
@@ -136,38 +164,160 @@ pub fn construct_tracer_payload(
+
+        let mut req = hyper::Request::builder()
+            .uri(target.url.clone())
+            .header(hyper::header::USER_AGENT, concat!("Tracer/", env!("CARGO_PKG_VERSION")))
```
what's this header used for?

bwoebi replied:
Telemetry and profiler use the header to clearly distinguish which code is the sender, in case some things are observed on the receiving side so that it's obvious what code sends it.

thedavl reviewed (commented):
Thanks for doing the re-factoring, I really like how you abstracted common logic. I left a couple questions regarding some stuff I wasn't clear on.

bwoebi commented on trace-utils/src/trace_utils.rs:
```diff
@@ -136,38 +164,160 @@ pub fn construct_tracer_payload(
+            };
+
+            send_request(req, serialized_trace_payload, StatusCode::ACCEPTED).await
+        } else {
```
Yes :-)
The Agent just takes individual collections of trace chunk, not a whole payload.

bwoebi commented on trace-utils/src/trace_utils.rs:
```diff
@@ -136,38 +164,160 @@ pub fn construct_tracer_payload(
-pub fn serialize_agent_payload(payload: pb::AgentPayload) -> anyhow::Result<Vec<u8>> {
+pub fn serialize_proto_payload<T>(payload: T) -> anyhow::Result<Vec<u8>>
+where
+    T: ::prost::Message,
```
Can be probably removed, I believe, I copied it from the generated protobuf files (where it's done to avoid conflicts...).

thedavl reviewed (dismissed):
looks good on the serverless end, thanks for the refactoring!

paullegranddc commented on ipc/src/platform/unix/mem_handle.rs:
```diff
@@ -0,0 +1,303 @@
+    }
+}
+
+pub trait FileBackedHandle<T>
+where
+    T: FileBackedHandle<T>,
```
If no, you can probably write it as `pub trait FileBackedHandle {`

bwoebi replied:
I need the T for `fn map(self) -> io::Result<MappedMem<T>>;` to have map return the right generic. Unless I'm missing something?

paullegranddc replied:
You can do `io::Result<MappedMem<Self>>` I think

paullegranddc replied:
Removing the generic parameter everywhere and add `where Self:Sized` to the trait works

bwoebi replied:
Ah, didn't know rust has `where Self:`. Thanks :-)

paullegranddc commented on ipc/src/platform/unix/mem_handle.rs:
```diff
@@ -0,0 +1,303 @@
+
+    pub fn open(path: String) -> io::Result<NamedShmHandle> {
+        let fd = shm_open(path.as_str(), OFlag::O_RDWR, Mode::empty())?;
+        let file: File = unsafe { OwnedFd::from_raw_fd(fd) }.into();
+        let size = file.metadata()?.size() as usize;
+        Self::new(file.into_raw_fd(), path, size)
```
My bad, I didn't know `into_raw_fd` took ownership if the file

bwoebi commented on ipc/macros/src/lib.rs:
```diff
@@ -0,0 +1,153 @@
+        }
+    }
+
+    camel_ty.shrink_to_fit();
```
This is macro-code though, so just compile time overhead.

bwoebi commented on ipc/src/transport/blocking.rs:
```diff
@@ -193,6 +193,10 @@ where
     }
 
+    pub fn is_closed(&self) -> bool {
+        self.transport.channel.probe_readable()
```
The blocking transport is not supposed to be readable on the client side unless it's a response. So, outside of waiting for a response, it will never be readable unless the server side closed its socket.

bwoebi commented on ipc/src/platform/unix/mem_handle.rs:
```diff
@@ -0,0 +1,303 @@
+
+fn page_aligned_size(size: usize) -> usize {
+    let page_size = page_size::get();
+    ((size - 1) & !(page_size - 1)) + page_size
```
That's confusing me now.. Isn't the `+ page_size` too much here, then?

bwoebi commented on ipc/src/platform/unix/mem_handle.rs:
```diff
@@ -0,0 +1,303 @@
+}
+
+struct ShmPath {
+    name: String,
```
Wasn't aware CString is a thing. Will try using it.

bwoebi commented on ipc/src/platform/unix/mem_handle.rs:
```diff
@@ -0,0 +1,303 @@
+    }
+}
+
+fn mmap_handle<T: FileBackedHandle<T>>(handle: T) -> io::Result<MappedMem<T>> {
+    let fd: RawFd = handle.get_shm().handle.as_raw_fd();
+    Ok(MappedMem {
+        ptr: unsafe {
+            mmap(
+                None,
+                NonZeroUsize::new(handle.get_shm().size).unwrap(),
+                ProtFlags::PROT_READ | ProtFlags::PROT_WRITE,
+                MapFlags::MAP_SHARED,
+                fd,
+                0,
+            )?
+        },
+        mem: handle,
+    })
+}
```
Which makes it public API of the trait though...

bwoebi commented on sidecar-ffi/src/unix.rs:
```diff
@@ -50,22 +59,139 @@ pub extern "C" fn ddog_ph_file_drop(ph: NativeFile) {
+{
+    let (new, contents) = reader.read();
+    // c_char may be u8 or i8 depending on target... convert it.
+    let contents: &[c_char] = unsafe { std::mem::transmute(contents) };
```
This doesn't work, because on mac it's i8, on linux u8, ... which is the reason for the transmute in the first place, instead of a cast :-X
See comment on the line above.

paullegranddc replied:
I'm not sure I understand `reader.read` always return a `&[u8]` and it's only the definition of c_char that changes right?

bwoebi commented on sidecar/src/agent_remote_config.rs:
```diff
@@ -0,0 +1,212 @@
+}
+
+fn path_for_endpoint(endpoint: &Endpoint) -> String {
+    let mut hasher = ZwoHasher::default();
```
Needs a stable hash, so that clients and server can find each other. I'll add a comment,

bwoebi commented on sidecar/src/interface.rs:
```diff
@@ -317,15 +373,199 @@ impl EnqueuedData {
+                    }
+                }
+
+                let mut data = self.inner.lock().unwrap();
+                let data = data.deref_mut();
+                if data.traces.send_data.is_empty() {
+                    data.flusher = None;
+                    break;
+                }
```
If there are no traces, the flusher will exit, right. But enqueue() (i.e. next received trace) will start a new flusher if there's none.

paullegranddc replied:
But why not just spawn it once and never exit?

bwoebi replied:
To avoid flushers being left-over from connections long closed.
So, it would require adding a synchronization point to notify about new data and adding a notification about the connection being closed. ... But that notification.

The current design was most natural for me to just have the flusher active when needed, then I don't need to worry about synchronizing with it manually.

paullegranddc commented on sidecar/src/agent_remote_config.rs:
```diff
@@ -0,0 +1,212 @@
+        data.meta.writing.store(true, Ordering::SeqCst);
+        data.meta.size = contents.len();
+
+        _ = data.as_slice_mut().write(contents);
```
If you use [copy_from_slice](https://doc.rust-lang.org/std/primitive.slice.html#method.copy_from_slice) you don't get the `Result` (although it will panic if the data slice is not big enough)
```suggestion
        data.as_slice_mut()[0..contents.length].copy_from_slice(contents);
```

paullegranddc commented on sidecar/src/agent_remote_config.rs:
```diff
@@ -0,0 +1,212 @@
+#[repr(C)]
+#[derive(Debug)]
+struct RawMetaData {
+    generation: u64,
```
Shouldn't this also be an atomic? It's read outside of section gated by `writing`

bwoebi replied:
Not sure if necessary, but can't do harm to make it atomic.

paullegranddc commented on sidecar/src/agent_remote_config.rs:
```diff
@@ -0,0 +1,212 @@
+    }
+}
+
+impl From<&[u8]> for &RawData {
+    fn from(value: &[u8]) -> Self {
+        // Safety: MappedMem is supposed to be big enough
+        unsafe { &*(value as *const [u8] as *const RawData) }
+    }
+}
```
I disagree with the safety of this operation.
First, in practice the slice is not guaranteed to have the right alignement, since you're copying data into a Vec, and then casting the underlying data to `&RawData` which can crash on ARM.
Second since this is a general trait, this can theoretically be called on any slice so the Safety comment is not really exact.

bwoebi replied:
In practice the Slice is mapped to the start of a page, so it should always be properly aligned, on a page of size >+ 4096.

paullegranddc replied:
You copy the page data to a Vec<u8> in the reader before casting it 

https://github.com/DataDog/libdatadog/blob/2cbbe1ee3318d00513c9e7bbc79180805f976c59/sidecar/src/agent_remote_config.rs#L143-L148

bwoebi replied:
So, what do you recommend here? I haven't seen any crashes any my M1 yet, but yeah... Is it possible to allocate an aligned Vec?

bwoebi replied:
Does it make sense to allocate a Vec<u64> instead? RawMetaData is repr(C), hence 64 bit aligned.
This should then make it safe, I assume?
And then reinterpret the &[u64] as &[u8] for returning.

paullegranddc replied:
I thin that instead of copying and keeping the whole memory in a contiguous slice, you can clone the sized RawMetaData to a field, and only the buf to a Vec.
Then you don't have to cast the whole vec to RawMetadata

paullegranddc commented on sidecar/src/agent_remote_config.rs:
```diff
@@ -0,0 +1,212 @@
+                    && new_generation == copied_data.meta.generation
+                {
+                    reader.current_config.replace(new_mem);
+                    return Some((true, new_data.as_slice()));
```
Here you return the slice pointing to the memory mapped region, but on line 171 you rerturn the copied Vec.

To prevent reads concurrent with writes, I think you should return the copied_data buffer instead

bwoebi commented on sidecar/src/config.rs:
```diff
@@ -111,13 +148,44 @@ impl FromEnv {
     }
 
+    fn self_telemetry() -> bool {
+        matches!(
+            std::env::var(ENV_SIDECAR_SELF_TELEMETRY)
+                .unwrap_or_default()
+                .as_str(),
+            "true" | "1"
+        )
```
```
error[E0308]: mismatched types
   --> /home/circleci/app/libdatadog/sidecar/src/config.rs:152:82
    |
152 |         matches!(std::env::var(ENV_SIDECAR_SELF_TELEMETRY).as_ref(), Ok("true" | "1"))
    |                  --------------------------------------------------              ^^^ expected struct `std::string::String`, found `str`
    |                  |
    |                  this expression has type `Result<&std::string::String, &VarError>`
    |
    = note: expected reference `&std::string::String`
               found reference `&'static str`
```
That doesn't seem to work.

paullegranddc replied:
Sorry, `as_deref()` instead of `as_ref()`

bwoebi commented on ddcommon-ffi/src/endpoint.rs:
```diff
@@ -0,0 +1,45 @@
+    let mut parts = Parts::default();
+    parts.authority = Authority::from_str(&unsafe { site.to_utf8_lossy() }).ok();
+    Box::new(Endpoint {
+        url: hyper::Uri::from_parts(parts).unwrap(),
```
It will panic when Authority::from_str() returns Err, which may happen if site contains certain special chars for example.
I'll change it to return MaybeError.

bwoebi commented:
@paullegranddc Thanks for the review. I've addressed the comments in the newest commit. Also rebased on main to fix conflicts.

The only remaining question is: this CI failure https://github.com/DataDog/libdatadog/actions/runs/5751516424/job/15590430623?pr=192
Do you have any idea how I work around that?

bwoebi commented:
I'm not even using the memfd thing from nix. Trying to disable with features seemingly did not help, even though it's the fs feature if I see that correctly, which isn't used?

bwoebi commented on sidecar/src/agent_remote_config.rs:
```diff
@@ -0,0 +1,224 @@
+
+                // Ensure the next write hasn't started yet *and* the data is from the expected generation
+                if !new_data.meta.writing.load(Ordering::SeqCst)
+                    && new_generation == copied_data.meta.generation.load(Ordering::Acquire)
```
The code was using the wrong data as source, needs to be new_data - the goal is checking whether the initial generation before I start copying is identical to the one after copying.
